"""knn homework.ipynb

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/drive/12iAwDFO1I1OATbQTZ1XlVgOavVbGm20r

Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.
"""

In [102]:
import math
import pandas as pd
import numpy as np

"""Make an untagged vector, point, and two tagged vectors, data1 and data2:"""

In [103]:

point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']

"""Write code to separate the data (X) from the tag (Y).  Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F

"""

In [104]:
print("The vector " , data1[0:-1], " has tag ", data1[-1])
print("The vector " , data2[0:-1], " has tag ", data2[-1])

The vector  [1, 1, 1]  has tag  M
The vector  [1, 3, 0]  has tag  F


"""Now let's classify the point as either M or F.  We'll do this by setting k = 1 and using the Euclidean distance.  We'll define that as:"""


In [105]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)

Manhattan Distance

In [106]:
def manhattanDistance(instance1, instance2, length):
    distance = 0
    for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance += abs(num1-num2)
    return distance

Hamming Distance

In [107]:
def hammingDistance(instance1, instance2, length):
    distance = 0
    for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance += int(num1 != num2) 
    return distance

Finds the closest k neighbors and takes mode

In [108]:
# multiple k
def knn(distances, k):
    # Assuming distances is a list of objects with a 'tag' attribute
    labels = [distance.tag for distance in distances[:k]]

    # Find the most common label
    label = max(set(labels), key=labels.count)
    return label

"""Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance."""


In [109]:
print("data1 " + str(euclideanDistance(data1,point,3)))
print("data2 " + str(euclideanDistance(data2,point,3)))

data1 1.4142135623730951
data2 3.0


"""Now, let's get more data from a file, myFile.csv. This code will read it for us:"""


In [110]:

url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0)
#put data in dataset without header line
dataset = np.array(df)

"""Show that you understand what you read by:

1. Printing the first two vectors in the file
2. Printing the Euclidean distance between those two vectors

Here is some code which I think might help get you on your way!
"""

In [111]:
print("first vector " + str(dataset[0]))
print("second vector " + str(dataset[1]))
print(euclideanDistance(dataset[0],dataset[1],3))

first vector [0 1 2 'F']
second vector [1 5 6 'F']
5.744562646538029


"""Now assume you have a new value for point:

[0,0,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:  

We suggest defining some type of data struction to store different vectors' distances and their tags like this:
"""

In [112]:
point=[0,0,100]
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point

"""You can then add vector distances like this:"""

In [113]:
eucDistances = [] # list of distances, will hold objects of type distClass
manDistances = []
hamDistances = []

for i in range(len(dataset)):
    temp=dataset[i]
    label=temp[-1]
    d = euclideanDistance(point,temp,3)
    m = manhattanDistance(point, temp, 3)
    h = hammingDistance(point, temp, 3)
    # print("The distances between " , point , " and " , temp,  " is " , str(d))
    # print(" and the label is " + label)
    eucObj = distClass() #one record's distance and tag
    eucObj.dist=d
    eucObj.tag=label
    eucDistances.append(eucObj)
    manObj = distClass()
    manObj.dist = m
    manObj.tag = label
    manDistances.append(manObj)
    hamObj = distClass()
    hamObj.dist = h
    hamObj.tag = label
    hamDistances.append(hamObj)

"""and sort them like this:"""

In [114]:
eucDistances.sort(key=lambda x: x.dist)
manDistances.sort(key=lambda x: x.dist)
hamDistances.sort(key=lambda x: x.dist)

"""Questions:

1. What is the label for point if k=1?
2. What is the label for point if k=3?
3. Would the result be different if we used a different distance function like Hamming or Manhattan?
"""

In [115]:
#Add code with functions that implement Hamming and Manhattan distances and test
#what the label will be for k=1 and k=3 for all possibilities
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)

print("euclidean")
print("for k = 1 label is " + knn(eucDistances, 1))
print("for k = 3 label is " + knn(eucDistances, 3))


print("manhattan")
print("for k = 1 label is " + knn(manDistances, 1))
print("for k = 3 label is " + knn(manDistances, 3))

print("hamming")
print("for k = 1 label is " + knn(hamDistances, 1))
print("for k = 3 label is " + knn(hamDistances, 3))

euclidean
for k = 1 label is M
for k = 3 label is M
manhattan
for k = 1 label is M
for k = 3 label is F
hamming
for k = 1 label is F
for k = 3 label is F


"""Up until here is a simplified version of the homework.
Below here is the work for the part we will be checking as the basis of your grade:

Now let's look at some bigger files:

1. mytrain.csv (for training the model)
2. mytest.csv (for testing)
"""

In [116]:
import numpy as np
import pandas as pd
import math

url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0))

print(train_data.shape)# number of records and features
print(train_data)

print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]
(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


"""I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:
1. k = 1
2. k = 7
3. k = 15

and at first use the Euclidean distance.
Classify each of the vectors in the test_data dataset using the training data from train_data.  Which value for k did the best?  What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.  

Once you are done, you should have a total of 9 different results:

1. Three results for the different value of k using the Euclidean Distance
2. Three results for the different value of k using the Hamming Distance
3. Three results for the different value of k using the Manhattan Distance

Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!
"""


SOLUTION

In [117]:
#Add code here

# instantiating variables for all accuracies
euc_accuracy_1 = 0
man_accuracy_1 = 0
ham_accuracy_1 = 0

euc_accuracy_7 = 0
man_accuracy_7 = 0
ham_accuracy_7 = 0

euc_accuracy_15 = 0
man_accuracy_15 = 0
ham_accuracy_15 = 0

# loop through test data
for i in test_data:
    eucDistances2 = [] # list of distances, will hold objects of type distClass
    manDistances2 = []
    hamDistances2 = []

    point = i
    point_label = point[-1]

    # for each point in test data we're going to find its distance to every point in training data
    # and add those distances to their respective lists
    for j in train_data: 
        temp = j
        label = temp[-1]
        length = len(temp) - 1

        d = euclideanDistance(point,temp, length)
        m = manhattanDistance(point, temp, length)
        h = hammingDistance(point, temp, length)

        eucObj = distClass() #one record's distance and tag
        eucObj.dist=d
        eucObj.tag=label
        eucDistances2.append(eucObj)

        manObj = distClass()
        manObj.dist = m
        manObj.tag = label
        manDistances2.append(manObj)

        hamObj = distClass()
        hamObj.dist = h
        hamObj.tag = label
        hamDistances2.append(hamObj)

    # sort the different lists that contain that distances from the test point to each train point
    eucDistances2.sort(key=lambda x: x.dist)
    manDistances2.sort(key=lambda x: x.dist)
    hamDistances2.sort(key=lambda x: x.dist)

    # check if result of knn is same as what tag is supposed to be 
    # k = 1
    if knn(eucDistances2, 1) == point_label:
        euc_accuracy_1 += 1
    if knn(manDistances2, 1) == point_label:
        man_accuracy_1 += 1
    if knn(hamDistances2, 1) == point_label:
        ham_accuracy_1 += 1
    # k = 7
    if knn(eucDistances2, 7) == point_label:
        euc_accuracy_7 += 1
    if knn(manDistances2, 7) == point_label:
        man_accuracy_7 += 1
    if knn(hamDistances2, 7) == point_label:
        ham_accuracy_7 += 1
    # k = 15
    if knn(eucDistances2, 15) == point_label:
        euc_accuracy_15 += 1
    if knn(manDistances2, 15) == point_label:
        man_accuracy_15 += 1
    if knn(hamDistances2, 15) == point_label:
        ham_accuracy_15 += 1


# extract the size of test data
size = len(test_data)

# divide all accuracies by size of test data to find their true accuracy

euc_accuracy_1 /= size
man_accuracy_1 /= size
ham_accuracy_1 /= size

euc_accuracy_7 /= size
man_accuracy_7 /= size
ham_accuracy_7 /= size

euc_accuracy_15 /= size
man_accuracy_15 /= size
ham_accuracy_15 /= size
    

RESULTS

In [118]:

euclidean_values = [euc_accuracy_1, euc_accuracy_7, euc_accuracy_15]  
manhattan_values = [man_accuracy_1, man_accuracy_7, man_accuracy_15]  
hamming_values = [ham_accuracy_1, ham_accuracy_7, ham_accuracy_15]  

# Create a DataFrame
data = {
    'k=1': [euclidean_values[0], manhattan_values[0], hamming_values[0]],
    'k=7': [euclidean_values[1], manhattan_values[1], hamming_values[1]],
    'k=15': [euclidean_values[2], manhattan_values[2], hamming_values[2]]
}

index = ['Euclidean', 'Manhattan', 'Hamming']

df = pd.DataFrame(data, index=index)
print(df)

            k=1   k=7  k=15
Euclidean  0.50  0.74  0.70
Manhattan  0.61  0.63  0.69
Hamming    0.61  0.55  0.57


Based on my results it seems the best value for k was:

k = 7 for Euclidean with accuracy 74%

k = 15 for Manhattan with accuracy 69%

k = 1 for Hamming with accuracy 61%